clients/python-wrapper/lakefs/object_io.py

```diff
@@ -0,0 +1,163 @@
+    This Object is instantiated and returned on open() methods for immutable reference types (Commit, Tag...)
+    """
+    _client: Client
+    _repo: str
+    _ref: str
+    _path: str
+    _pos: int
+    _pre_sign: bool
+
+    def __init__(self, repository: str, reference: str, path: str,
+                 pre_sign: Optional[bool] = None, client: Optional[Client] = DefaultClient) -> None:
+        self._client = client
+        self._repo = repository
+        self._ref = reference
+        self._path = path
+        self._pre_sign = pre_sign if pre_sign is not None else client.storage_config.pre_sign_support
+        self._pos = 0
+
+    @property
+    def pos(self):
+        return self._pos
+
+    def seek(self, pos):
+        if pos < 0:
+            raise ValueError("position must be a non-negative integer")
+        self._pos = pos
+
+    def read(self, read_bytes: int = None) -> bytes:
+        if read_bytes and read_bytes <= 0:
+            raise ValueError("read_bytes must be a positive integer")
+        try:
+            stat = self._client.sdk_client.objects_api.stat_object(self._repo, self._ref, self._path)
+        except NotFoundException:
+            raise ObjectNotFoundException
+        if self._pos >= stat.size_bytes:
+            raise EOFError
+        read_bytes = read_bytes if read_bytes is not None else stat.size_bytes
+        new_pos = min(self._pos + read_bytes, stat.size_bytes)
+        read_range = _RANGE_STR_TMPL.format(start=self._pos, end=new_pos - 1)
+        contents = self._client.sdk_client.objects_api.get_object(self._repo,
+                                                                  self._ref,
+                                                                  self._path,
+                                                                  range=read_range,
+                                                                  presign=self._pre_sign)
+        self._pos = new_pos  # Update pointer position
+        return contents
+
+    def stat(self):
+        return self._client.sdk_client.objects_api.stat_object(self._repo, self._ref, self._path)
+
+    def exists(self):
+        try:
+            self._client.sdk_client.objects_api.head_object(self._repo, self._ref, self._path)
+            return True
+        except NotFoundException:
+            return False
+
+
+class WriteableObject(ReadableObject):
+    """
+    WriteableObject inherits from ReadableObject and provides read/write functionality for lakeFS objects 
+    using IO semantics.
+    This Object is instantiated and returned upon invoking open() on Branch reference type.
+    """
+
+    def __init__(self, repository: str, reference: str, path: str,
+                 pre_sign: Optional[bool] = None, client: Optional[Client] = DefaultClient) -> None:
+        # Verify that reference is a branch, otherwise throws exception
+        try:
+            client.sdk_client.branches_api.get_branch(repository, reference)
```

Comment:
IO in a constructor should probably be avoided (i.e. how would you handle listing on a branch which should return a generator of `WriteableObject`?

Reply:
removed
This was a bad idea though good intentions 😄